### ISSUE TITLE
Unable to clear UART interrupts once they have occured

### ISSUE THREAD

**Issue by alistair23** [lowRISC/opentitan#1232]
The issue that I'm seeing is that once an UART interrupt has occurred, it can't be cleared.

That is, once an interrupt from the UART device has triggered (`intr_state` register is set), the `intr_state` register can't be cleared and I don't see any more interrupts from the UART device.

### Steps to reproduce

You will need to clone the OT software from here: https://github.com/alistair23/opentitan/tree/alistair/uart-int-wip

This is based on the [UART interrupt work](https://github.com/silvestrst/opentitan/tree/silvestrs/uart-dif-prototype-interrupts-2) by @silvestrst, but has two extra commits on top.

The provided code can be run in verilator (interrupts don't work on the FPGA, still investigating) using the normal steps.

I run something like this:

```shell
./meson_init.sh
fusesoc --cores-root . run --target=sim --setup --build lowrisc:systems:top_earlgrey_verilator
ninja -C build-out/sw/sim-verilator/ all
uild/lowrisc_systems_top_earlgrey_verilator_0.1/sim-verilator/Vtop_earlgrey_verilator   --meminit=rom,build-out/sw/sim-verilator/sw/device/boot_rom/boot_rom.elf   --meminit=flash,build-bin/sw/device/sim-verilator/examples/hello_world/hello_world.elf
```

I then quickly connect to the serial prompt using screen

```shell
screen /dev/pts/11
```

You will see something like this, make sure you don't press any buttons yet:

```
Build Date: 2019-12-18, 13:44:57
INFO: Boot ROM initialisation haExternal IRQ triggered!
                                                       TX WTAERMARK IRQ!
                                                                        handler_uart_isr: implement tx watermark!
                                                                                                                 Hello World! Dec 18 2019 23:57:12
Watch the LEDs!
Try out the switches on the board
or type anything into the console window.
The LEDs show the ASCII code of the last character.

```

Note that `External IRQ triggered!`, `TX WTAERMARK IRQ!` and `handler_uart_isr: implement tx watermark!` means that the TX watermark (set to 0) IRQ occurred. The SW has handled the interrupt and cleared it from the `PLIC` and `MIP`

If you read the UART `intr_state` register `0x40000000` you will see a value of `0x00000001`.

NOTE, I use GDB to read it, with these steps:
Start OpenOCD:

```shell
openocd -s util/openocd -f board/lowrisc-earlgrey-verilator.cfg
```

Run GDB

```shell
riscv32-oe-elf-gdb /scratch/alistair/software/tier2/opentitan/build-out/sw/sim-verilator/sw/device/boot_rom/boot_rom.elf
target extended-remote :3333
x/x 0x40000000
```

Back in the screen serial session if you enter a charecter (you can do it at any time) you will see this:

```
External IRQ triggered!
                       received char
```

That means the RX watermark interrupt occurred, if you read `intr_state` after this it will be `0x00000003`.

This [commit](https://github.com/alistair23/opentitan/commit/2431f5abe870420e5e1f6375d35e7097e5b8fe36) in the repo tries to clear these interrupts after they have occured and it unable to.

I have tried writing 0 to the UART `intr_state`, UART `fifo_ctrl` (for `rxrst` and `txrst`)  and even the PLIC `cc0` but nothing clears the UART `intr_state` and I am unable to repeat the RX or TX water mark interrupts.

I don't see any other ways to clear the UART interrupts from the UART device.

**Comment by tjaychen** [comment]
@alistair23 
the tx/rx watermark interrupts right now are level driven... meaning if your uart FIFO depth is more than or less than your programmed threshold, the interrupt will keep on asserting even after you clear it.

Do you think you'd prefer a more edge behavior? (there are a few other places where we've made similar changes).  So the basic idea in this case is that you'd have to fill in more data, or read out of the data (in case of rx) for the event to go away. 

Regarding `rxtrst` and `txrst`, those are supposed to be active high signals.  Do you mean you programmed them to 1 and it did not work?

**Comment by alistair23** [comment]
Let's ignore TX watermark for one second, as there is this open issue: https://github.com/lowRISC/opentitan/issues/1132

For the RX case, this is what I image.

 1. SW sets RX watermark to 0, it wants an interrupt anytime there is a char received
 2. The interrupt occurs, SW reacts, reads the data or something else and then clears that interrupt
 3. SW goes back to waiting for the next char

In this case SW should be able to clear the interrupt, while leaving the watermark where it was.

Right now this is especially confusing as the UART keeps "asserting" an interrupt, but the PLIC doesn't react to it.

> So the basic idea in this case is that you'd have to fill in more data, or read out of the data (in case of rx) for the event to go away.

That works for me, the problem is that the interrupt doesn't go away after reading the data.

> Regarding `rxtrst` and `txrst`, those are supposed to be active high signals. Do you mean you programmed them to 1 and it did not work?

Ah, whoops. I only programmed them to 0. That's my mistake. I tried setting it to 3 (reset `TX` and `RX`) and it messes with the printing a bit (as I'm resetting the UART) but it also doesn't clear the `intr_state` register.

**Comment by alistair23** [comment]
I just pushed an update to my example OT repo mentioned above. The new commit that tries to clear the `intr_state` register is here: https://github.com/alistair23/opentitan/commit/4df4e97999d57ea8967f974bdddd141b3db008ae

It's more aggressive as it reads the rdata buf a few times and now correctly asserts `rxrst` and `txtst` but the `intr_state` bits and still not cleared.

**Comment by alistair23** [comment]
I have tried writing 1 to the `intr_state` register but that doesn't clear them either. I tried writing `0xFFFF_FFFF` but that just resulted in the register being set to `0xb` (`tx_watermark`, `rx_watermark` and somehow `rx_overflow`..).

I tried setting it back to 0 after setting it to 1 and that results in the same value of `0x3` (`tx_watermark` and `rx_watermark`).

**Comment by silvestrst** [comment]
@tjaychen am understanding your comment correctly - TX and RX watermarks will assert until the respective FIFOs are drained above/below the watermark?

How I imagined this works was that the interrupt is triggered on transition, for example:
RX watermark is set to 4 bytes, FIFO grows to 16 bytes (just for the argument sake) - the interrupt is triggered only once, even if SW doesn't drain RX FIFO below 4 bytes. I think that was @alistair23 understanding as well, but please correct me if I am wrong.

**Comment by tjaychen** [comment]
yes your understanding is correct. Eunchan please correct me if I am wrong.
There have been a few places in our code base where we've moved to edge,
could this be one of those as well?

I guess my worry for an edge usage would be something like the following:

- we set a threshold x
- the FIFO receives more than x, and an interrupt is generated
- software comes in, notices there are y entries (because more have been
deposited during the interrupt switch)
- software reads out y entries, however during this time, even more entries
are deposited, the end result being we are still above x
- we exit interrupt handling, a new interrupt is not generated since we
never dipped below x.

In the above example, you'd need to have some kind of check at the end of
handling to ensure you at least drained below x to make sure another
interrupt comes.  Whereas with level you would not.
Both could work, so I guess it's really a matter of which we think is
better?

**Comment by tjaychen** [comment]
@alistair23 
i just cloned your repo and i'm giving it a try.  One thing I noticed is that in `handler.c` the clearing of `CC0` doesn't seem to be right. 

Our documentation requires us to write the irq_data back into cc0 in order to "complete" the handling.
I tried that in my local copy of yours, and now i'm getting a storm of tx_watermark interrupts, which is really related to #1132.

I'll quickly fix that locally and see how it goes.

**Comment by alistair23** [comment]
Ah ha!

Writing the claim register back to itself does seem to allow multiple interrupts.

I don't see anything in the PLIC documentation saying that you have to write the exact value back: https://github.com/riscv/riscv-plic-spec/blob/master/riscv-plic.adoc#interrupt-claim-process

**Comment by alistair23** [comment]
Ah, it's here: https://github.com/riscv/riscv-plic-spec/blob/master/riscv-plic.adoc#interrupt-completion

I see, I missed that part.

**Comment by tjaychen** [comment]
indeed! now i'm getting double rx interrupts when i type into the console. I'm guessing this is related to how we clear it locally, but unfortunately i have to run.

I will have another look tonight.

**Comment by alistair23** [comment]
On Verilator I am also seeing double RX interrupts.

On the FPGA I still don't see any though.

Thanks for your help.

Has anyone tested interrupts on the FPGA?

**Comment by tjaychen** [comment]
I've only tried the timer one, but that was purely using the peripheral and
not plic (so it was one and done).

I'll give that a try too once we get through the verilator issue. In the
fpga we probably need to check if the uart state is even filling.

**Comment by tjaychen** [comment]
O one difference between verilator and fpga. Fpga actually has pin
configutations, and verilator does not right now (we want to change this in
the future). So maybe that one isn't entirely correct on the rx pin, that
could explain why you're not seeing anything.

**Comment by alistair23** [comment]
I fixed the double interrupt issue, that was just a mistake in when we complete the interrupt.

So on Verilator everything seems to be working, thanks so much for the help.

Still no luck on the FPGA.

What do you mean isn't correct? I see the UART data being printed, just not an interrupt.

**Comment by alistair23** [comment]
On the FPGA I see the PLIC `pending` register being set, but nothing in the `claim` register and no interrupt.

**Comment by tjaychen** [comment]
Just to clarify, is the pending bit being set the right one? Im actually
not sure if we expose the mapping in a clear way.

Also, the fpga you are using, is it one of the released versions or did you
build it directly?

**Comment by alistair23** [comment]
I think the `pending` register value is `0x02`.

I'm using the one built by the CI: opentitan-snapshot-20191101-1-384-gc4d8c97

The released one has issues with debuggers that have been fixed since then so it isn't reliable for debugging.

**Comment by tjaychen** [comment]
ah did you read RV_PLIC_IP0 or 1? I'm guessing 1?
Also, it might be good to double check the interrupt state inside uart to
make sure it has actually asserted, that way we can rule out it's a
pad->uart issue, and focus on whatever the problem is inside plic.

The other part is that your fpga binary is likely before this change (
https://github.com/lowRISC/opentitan/commit/dddaa14a8559359189ba1615aa95ee46ba8fcad6#diff-7f62678b832d21be9f19de52c63ef25a
).
So we should probably get you the latest binary (one that mostly matches
your verilator build) to try.

**Comment by imphil** [comment]
Just FYI, you can get up-to-date FPGA bitstreams from Azure Pipelines. Go to https://dev.azure.com/lowrisc/opentitan/_build?definitionId=5&_a=summary&repositoryFilter=5&branchFilter=566, choose a recent build, and click on the top-right on "Artifacts" (e.g. https://dev.azure.com/lowrisc/opentitan/_build/results?buildId=6745&view=artifacts&type=publishedArtifacts). Download opentitan-dist for a binary snapshot which includes a bitstream.

**Comment by tjaychen** [comment]
thanks for the note @imphil 
Do you think we should put this information somewhere in the user docs? I'm sure there will be places where an active developer wants to debug something on the latest build and not one of the released builds.

**Comment by alistair23** [comment]
I just updated to: `opentitan-snapshot-20191101-1-399-g15e1026`, built from commit `15e1026ec5e231be00c1599ade5090c9915b0e5e`, which is the latest I can download without a Google account.

Still no interrupts on the FPGA.

@tjaychen I was reading `RV_PLIC_IP1`. I can see that the interrupts state inside the UART is asserted

> Do you think we should put this information somewhere in the user docs? I'm sure there will be places where an active developer wants to debug something on the latest build and not one of the released builds.

I think it would be better to spend that time releasing more often, as lots of people will still try to use release builds.

**Comment by alistair23** [comment]
Just confirming with more values:

On the FPGA:
`RV_PLIC_IP1` -> `0x00000002` after a RX IRQ
Nothing in the PLIC `claim` register

**Comment by tjaychen** [comment]
@alistair23 
sorry i've been meaning to attach this but keep forgetting.  I pulled your repo and just built a new fpga bitfile. Can you try replacing your current bitfile with this one and see how it goes? 

[lowrisc_systems_top_earlgrey_nexysvideo_0.1.bit.tar.gz](https://github.com/lowRISC/opentitan/files/4006347/lowrisc_systems_top_earlgrey_nexysvideo_0.1.bit.tar.gz)

**Comment by tjaychen** [comment]
also to confirm, i'm guessing all the interrupt enable, threshold settings are the same between your fpga and verilator correct?

**Comment by alistair23** [comment]
I just tried that bitstream and it still doesn't look like it's working. Do interrupts work for you?

**Comment by alistair23** [comment]
Yeah, I'm running the same source code on both

**Comment by tjaychen** [comment]
I'm not next to an FPGA right now so I can't really try :(
The fact that it's not working for you isn't really a good sign...

Just to make sure we're on the same page..basically you're saying the
interrupt state sets in uart, the IP bit also sets in PLIC, but you don't
see anything to claim, is that right?

**Comment by alistair23** [comment]
Yes, that is correct. I see all of that while stopped (with GDB) so nothing else can claim the interrupt.

**Comment by tjaychen** [comment]
could you just dump out all the rv_plic registers at this point?

**Comment by eunchan** [comment]
> could you just dump out all the rv_plic registers at this point?

And also UART register too.

**Comment by alistair23** [comment]
Here are the UART registers:
```
0x40000000:	0x00000003
0x40000004:	0x00000002
0x40000008:	0x00000000
0x4000000c:	0x12df0003
0x40000010:	0x0000001c
0x40000014:	0x0000000d
0x40000018:	0x00000000
0x4000001c:	0x00000000
0x40000020:	0x00000000
0x40000024:	0x00000000
0x40000028:	0x0000ffff
0x4000002c:	0x00000000
0x40000030:	0xffffffff
```

and the PLIC

```
0x40090000:	0x00000000
0x40090004:	0x00000002
0x40090008:	0x00000000
0x4009000c:	0x00000000
0x40090010:	0x00000000
0x40090014:	0x00000000
0x40090018:	0x00000000
0x4009001c:	0x00000000
0x40090020:	0x00000000
0x40090024:	0x00000000
0x40090028:	0x00000000
0x4009002c:	0x00000000
0x40090030:	0x00000000
0x40090034:	0x00000000
0x40090038:	0x00000000
0x4009003c:	0x00000000
0x40090040:	0x00000000
0x40090044:	0x00000000
0x40090048:	0x00000000
0x4009004c:	0x00000000
0x40090050:	0x00000000
0x40090054:	0x00000000
0x40090058:	0x00000000
0x4009005c:	0x00000000
0x40090060:	0x00000000
0x40090064:	0x00000000
0x40090068:	0x00000000
0x4009006c:	0x00000000
0x40090070:	0x00000000
0x40090074:	0x00000000
0x40090078:	0x00000000
0x4009007c:	0x00000000
0x40090080:	0x00000000
0x40090084:	0x00000000
0x40090088:	0x00000000
0x4009008c:	0x00000000
0x40090090:	0x00000003
0x40090094:	0x00000003
0x40090098:	0x00000000
0x4009009c:	0x00000000
0x400900a0:	0x00000000
0x400900a4:	0x00000000
0x400900a8:	0x00000000
0x400900ac:	0x00000000
0x400900b0:	0x00000000
0x400900b4:	0x00000000
0x400900b8:	0x00000000
0x400900bc:	0x00000000
0x400900c0:	0x00000000
0x400900c4:	0x00000000
0x400900c8:	0x00000000
0x400900cc:	0x00000000
0x400900d0:	0x00000000
0x400900d4:	0x00000000
0x400900d8:	0x00000000
0x400900dc:	0x00000000
0x400900e0:	0x00000000
0x400900e4:	0x00000000
0x400900e8:	0x00000000
0x400900ec:	0x00000000
0x400900f0:	0x00000000
0x400900f4:	0x00000000
0x400900f8:	0x00000000
0x400900fc:	0x00000000
0x40090100:	0x00000000
0x40090104:	0x00000000
0x40090108:	0x00000000
0x4009010c:	0xffffffff
0x40090110:	0xffffffff
0x40090114:	0xffffffff
0x40090118:	0xffffffff
0x4009011c:	0xffffffff
0x40090120:	0xffffffff
0x40090124:	0xffffffff
0x40090128:	0xffffffff
0x4009012c:	0xffffffff
0x40090130:	0xffffffff
0x40090134:	0xffffffff
0x40090138:	0xffffffff
0x4009013c:	0xffffffff
0x40090140:	0xffffffff
0x40090144:	0xffffffff
0x40090148:	0xffffffff
0x4009014c:	0xffffffff
0x40090150:	0xffffffff
0x40090154:	0xffffffff
0x40090158:	0xffffffff
0x4009015c:	0xffffffff
0x40090160:	0xffffffff
0x40090164:	0xffffffff
0x40090168:	0xffffffff
0x4009016c:	0xffffffff
0x40090170:	0xffffffff
0x40090174:	0xffffffff
0x40090178:	0xffffffff
0x4009017c:	0xffffffff
0x40090180:	0xffffffff
0x40090184:	0xffffffff
0x40090188:	0xffffffff
0x4009018c:	0xffffffff
0x40090190:	0xffffffff
0x40090194:	0xffffffff
0x40090198:	0xffffffff
0x4009019c:	0xffffffff
0x400901a0:	0xffffffff
0x400901a4:	0xffffffff
0x400901a8:	0xffffffff
0x400901ac:	0xffffffff
0x400901b0:	0xffffffff
0x400901b4:	0xffffffff
0x400901b8:	0xffffffff
0x400901bc:	0xffffffff
0x400901c0:	0xffffffff
0x400901c4:	0xffffffff
0x400901c8:	0xffffffff
0x400901cc:	0xffffffff
0x400901d0:	0xffffffff
0x400901d4:	0xffffffff
0x400901d8:	0xffffffff
0x400901dc:	0xffffffff
0x400901e0:	0xffffffff
0x400901e4:	0xffffffff
0x400901e8:	0xffffffff
0x400901ec:	0xffffffff
0x400901f0:	0xffffffff
0x400901f4:	0xffffffff
0x400901f8:	0xffffffff
0x400901fc:	0xffffffff
0x40090200:	0x00000000
0x40090204:	0x00000003
0x40090208:	0x00000001
0x4009020c:	0x00000000
0x40090210:	0x00000000
0x40090214:	0xffffffff
0x40090218:	0xffffffff
0x4009021c:	0xffffffff
0x40090220:	0xffffffff
```

This is after the hello_world binary has run to the loop and I just entered some text (which should generate an interrupt).

**Comment by tjaychen** [comment]
i glanced quickly at these, nothing really stands out to me right now...
I'd probably start experimenting with setting different enables / priorities / thresholds just to see if there's anything that alters the behavior (i could try that next week also when i'm actually in front of an FPGA).

Just to be clear also..these results, are they using the released fpga binfile? O are you using the one I attached?

**Comment by alistair23** [comment]
They are using the bitstream you shared. Although the one built from the latest commit that doesn't require a Google account also doesn't work. I'm assuming it's the same values.

**Comment by tjaychen** [comment]
thanks alistair.
@eunchan do you see anything that stands out to you from the reg dumps?  I didn't really see anything.

**Comment by pfmooney** [comment]
I've been working on driving the UART with a modified version of what @alistair23 [drafted up for Tock](https://github.com/tock/tock/pull/1460).  Using a verilator model built from OpenTitan as of yesterday (including the changes for tx_empty interrupts on the UART), I've found that the internal interrupt state changes do not appear to make sense.

In particular, I was filling the TX FIFO, and then enabling the tx_empty UART to receive notification when it drained.  With a full trace from the verilator model, I can see some of the signals in the UART which to me seem confusing:
![2020-01-08-093124_2590x380_scrot](https://user-images.githubusercontent.com/1555070/71997710-f41c2f80-3203-11ea-8d6a-eeb457466054.png)

The `tx_empty_d` signal matches up with the fifo level starting at (and temporarily) dropping down to 0.  When that is observed by the tx_empty reg2hw block, it does seem to trigger what would have been an interrupt condition, but the system software has not yet set the `tx_empty` bit in the `intr_enable` register.  Once the software has filled the buffer, and does set the bit to enable that interrupt, the `intr_o` output strangely goes high, even though the FIFO is clearly at a non-zero depth.

Am I misunderstanding the situation here?  Are there other piece of information which would be useful to collect?  I can share the signal trace from verilator if that would help.

**Comment by tjaychen** [comment]
@pfmooney thanks for input!
I think you are also seeing a bit of a side effect, please see #1290 

There are a couple of reasons why you're seeing this.  

The TX_EMPTY interrupt is right now an edge event.  Meaning as long as it has happened once, the interrupt state will latch and wait for software to clear it. 

The other issue is that interrupt enable does not actually prevent the interrupt state from setting, it simply controls whether or not the interrupt is actually shown to the PLIC (which has yet another level of enable).

**Comment by eunchan** [comment]
Assuming the bitstream was before #1290 then I don't see any particular reason why tx watermark interrupt cannot be cleared. tx_fifo_depth is 0 then previous RTL should not create the tx_watermark event. 

Is it still reproducible with latest design?

**Comment by alistair23** [comment]
@eunchan the interrupt can be cleared (the title in this issue is out of date). The main issue now is that interrupts don't work in the FPGA. I didn't create a new issue as there is a fair bit of relevant discussion here, but if you want I'm happy to close this and create a new one.

**Comment by eunchan** [comment]
Thanks for update. Let me go through the comments above to figure out how the software/ hardware are configured.

**Comment by eunchan** [comment]
@tjaychen I think we have CDC issue now... don't we? (UART uses peri clock and the PLIC uses main clock)

**Comment by tjaychen** [comment]
yes technically we would, but all the separate clocks are still tied to the
same source in both FPGA / design isn't it?
we have not actually split it out into multiple sources yet (partially
because i haven't really had a chance to look into how this might impact
the verilator DPI modules)

**Comment by asb** [comment]
Now #1364 has been merged, can this be closed?

**Comment by alistair23** [comment]
Closing as this has all been fixed
